### ISSUE TITLE
Waiting for Shipyard on x.x.x.x:8080

### ISSUE THREAD

**Issue by John0x** [shipyard/shipyard#897]
Hey, I've tried to install shipyard on a fresh Ubuntu 14.04 LTS system (after installing the docker cli according to offial docs).
It loads the containers but upon starting it just hangs forever.

The firewall is inactive.

````
Digest: sha256:5f065362680fa4565dd150c8da3edd09b79a7a3010d3ceef20093c2a879187e0
Status: Downloaded newer image for shipyard/shipyard:latest
Waiting for Shipyard on x.x.x.x:8080
...........................................................................................................................
````

**Comment by ehazlett** [comment]
That doesn't look like an internal address.  Can you check if there is a
firewall and ensure port 8080 is allowed?  Although if you are going to use
it remotely you will want to configure TLS or it will be accessible to the
public.

**Comment by xhz-bza** [comment]
Can you reply to me? #894 this problem @ehazlett

**Comment by xhz-bza** [comment]
Is also the problem on Ubuntu 14.04 LTS system，ufw Already allow 8080/tcp

check the log,the following output

INFO[0000] shipyard version 3.1.0                       
DEBU[0000] using tls for communication with docker      
DEBU[0000] setting up rethinkdb session                 
WARN[0000] Error creating connection: gorethink: dial tcp 113.12.83.4:28015: getsockopt: connection refused 
FATA[0000] no connections were made when creating the session

Where is the IP address from？

I use the 'curl -sSL https://shipyard-project.com/deploy | bash -s'

**Comment by John0x** [comment]
Ah sorry, I replaced the ip with xxxx, it's just a typical internal ip.
There firewall ufw is disabled.

**Comment by xhz-bza** [comment]
How do you solve？
check the log,the following output

INFO[0000] shipyard version 3.1.0

DEBU[0000] using tls for communication with docker

DEBU[0000] setting up rethinkdb session

WARN[0000] Error creating connection: gorethink: dial tcp 113.12.83.4:28015: getsockopt: connection refused
FATA[0000] no connections were made when creating the session

I use the 'curl -sSL https://shipyard-project.com/deploy | bash -s'

**Comment by dguerizec** [comment]
Same problem here :(

**Comment by ehazlett** [comment]
What do `docker logs shipyard-controller` report?

**Comment by GewoonMaarten** [comment]
`docker logs shipyard controller` says

`INFO[0000] shipyard version 3.1.0                       
DEBU[0000] setting up rethinkdb session                 
WARN[0000] Error creating connection: gorethink: dial tcp 127.0.53.53:28015: getsockopt: connection refused 
FATA[0000] no connections were made when creating the session`

I used the automated deployment

**Comment by luca-moser** [comment]
Same here using: 
* `curl -sSL https://shipyard-project.com/deploy | ACTION=upgrade PORT=8484 IP=0.0.0.0 bash -s`
* `curl -sSL https://shipyard-project.com/deploy | ACTION=upgrade PORT=8080 IP=127.0.0.1 bash -s`

Output:
```
Status: Downloaded newer image for shipyard/shipyard:latest
 -> Upgrading Controller
Waiting for Shipyard on 127.0.0.1:8080
......................................................................................................................................................................................................................
```

**Comment by ehazlett** [comment]
Can you post your logs from the shipyard-controller container?

**Comment by sergey789** [comment]
If the one does this on ARM platform - same issue with eternal "Waiting for Shipyard". 
Fix: docker image is not suitable for ARM, use ARM specific shipyard docker image.

**Comment by luca-moser** [comment]
Isn't the problem the bash script itself? The `curl container` tries to curl the ip of the `shipyard-controller container` but since the shipyard-controller is running on `127.0.0.1` , the `curl container` will curl itself on what ever port we defined, never getting any response:

```
wait_for_available() {
    set +e 
    IP=$1
    PORT=$2
    echo Waiting for Shipyard on $IP:$PORT

    docker pull ehazlett/curl > /dev/null 2>&1

    TLS_OPTS=""
    if [ $ENABLE_TLS = 1 ]; then
        TLS_OPTS="-k"
    fi

    until $(docker run --rm ehazlett/curl --output /dev/null --connect-timeout 1 --silent --head --fail $TLS_OPTS $SHIPYARD_PROTOCOL://$IP:$PORT/ > /dev/null 2>&1); do
        printf '.'
        sleep 1 
    done
    printf '\n'
}
```

**Comment by Kizoku-Dev** [comment]
Same problem for me.

Do you have any news about it ?

**Comment by jos666** [comment]
Me too
shipyard-controller  Infinite restart
docker logs:
```
INFO[0000] shipyard version 3.1.0                       
WARN[0000] Error creating connection: gorethink: dial tcp 98.124.199.16:28015: getsockopt: connection refused 
FATA[0000] no connections were made when creating the session 
INFO[0000] shipyard version 3.1.0
```
Installation is: 
```
curl -sSL https://shipyard-project.com/deploy | bash -s
```
Container can access rethinkdb inside, may be fixed in the program rethinkdb domain name?

**Comment by guppie70** [comment]
Same problem here. 
Host machine: macOS Sierra (10.12.2)

Installed Shipyard using: `curl -sSL https://shipyard-project.com/deploy | bash -s`
After installation, the script kept waiting for port 8080.

Restarted my machine - no response on http://127.0.0.1:8080

Log of the shipyard-controller shows:
```
INFO[0000] shipyard version 3.1.0                       
DEBU[0000] setting up rethinkdb session                 
WARN[0022] Error creating connection: gorethink: dial tcp 172.17.0.2:28015: getsockopt: connection refused 
```

Has anybody solved this problem?

**Comment by guppie70** [comment]
I have solved it!
After doing some more research, I found this post: https://github.com/shipyard/shipyard/issues/880
It suggested to add a file /etc/nsswitch.conf in the container. Not being a Linux expert, I copied the file from an AWS EC2 instance which looks like this:
```
#
# /etc/nsswitch.conf
#
# An example Name Service Switch config file. This file should be
# sorted with the most-used services at the beginning.
#
# The entry '[NOTFOUND=return]' means that the search for an
# entry should stop if the search in the previous entry turned
# up nothing. Note that if the search failed due to some other reason
# (like no NIS server responding) then the search continues with the
# next entry.
#
# Valid entries include:
#
#	nisplus			Use NIS+ (NIS version 3)
#	nis			Use NIS (NIS version 2), also called YP
#	dns			Use DNS (Domain Name Service)
#	files			Use the local files
#	db			Use the local database (.db) files
#	compat			Use NIS on compat mode
#	hesiod			Use Hesiod for user lookups
#	[NOTFOUND=return]	Stop searching if not found so far
#

# To use db, put the "db" in front of "files" for entries you want to be
# looked up first in the databases
#
# Example:
#passwd:    db files nisplus nis
#shadow:    db files nisplus nis
#group:     db files nisplus nis

passwd:     files sss
shadow:     files sss
group:      files sss

#hosts:     db files nisplus nis dns
hosts:      files dns

# Example - obey only what nisplus tells us...
#services:   nisplus [NOTFOUND=return] files
#networks:   nisplus [NOTFOUND=return] files
#protocols:  nisplus [NOTFOUND=return] files
#rpc:        nisplus [NOTFOUND=return] files
#ethers:     nisplus [NOTFOUND=return] files
#netmasks:   nisplus [NOTFOUND=return] files     

bootparams: nisplus [NOTFOUND=return] files

ethers:     files
netmasks:   files
networks:   files
protocols:  files
rpc:        files
services:   files sss

netgroup:   files sss

publickey:  nisplus

automount:  files
aliases:    files nisplus

```

Then, on my running container I used 
`docker exec -it shipyard-controller sh`
to access my container and then
`cd /etc`
then
```
cat <<'EOF' > nsswitch.conf
--- all the naswitch.conf file content
EOF
```

And voila: http://localhost:8080 is working!

But - now my question: how can I add this file to my container so that the next time I boot my machine and Docker starts up the containers again, I do not need to repeat the above?
Anybody knows how to achieve this?

**Comment by david9991** [comment]
see: `docker commit` command.
After you add the nsswitch.conf to your container, you can commit this change as a new docker image which for future you use it.

**Comment by ktiedt** [comment]
This is also needed for ubuntu 16.04 it seems :(

**Comment by ivychill** [comment]
@guppie70 How can you run "docker exec -it shipyard-controller sh", given that shipyard-controller is restarting Infinitely.
I get "Error response from daemon: Container  ... is restarting, wait until the container is running" when i run "docker exec -it shipyard-controller sh"

**Comment by ktiedt** [comment]
@ivychill they described how they worked around that problem in the comment just above... for simplicity sake here is the link to that comment https://github.com/shipyard/shipyard/issues/897#issuecomment-269032252

**Comment by ivychill** [comment]
@ktiedt Thank you! I have read that, but my problem is that I cannot access the container, since the container "shipyard-controller" is restarting Infinitely.

**Comment by ehazlett** [comment]
@guppie70 @ktiedt can you give me a little more info on your host setup?  I can add this to the image but I'm reluctant as I haven't been able to reproduce and would rather find the root cause.

**Comment by ktiedt** [comment]
@ehazlett well in my case its a pretty standard Ubuntu 16.04 server install running latest docker/docker-compose.

**Comment by david9991** [comment]
@ivychill By `docker cp` command. You can copy any file to container without it got be run. You can copy nsswitch.conf to container directly, or overwrite 'entrypoint' script to prevent a container from infinite restart.

**Comment by mariolameiras** [comment]
The /etc/nslookup.conf workaround solved it for me as well with Docker version 17.03.0-ce and ubuntu-server 16.04

**Comment by ehazlett** [comment]
Can you also check what `/etc/hosts` is before adding the file?  I want to see if something is getting improperly added.
